### ISSUE TITLE
Make initial notification in ONCHANGE notifications configurable

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#920]
Currently, notification is always sent. A URI parameter should be defined to select wether the notification has to be sent or not.

**Comment by fgalan** [comment]
The following answers at SOF should be updated once this get implemented:

* http://stackoverflow.com/questions/30195071/old-notifications-pushed-immediately-when-subscription-is-created
* http://stackoverflow.com/questions/32370847/subscribe-orion-without-receive-old-entities/

**Comment by fgalan** [comment]
As alternative (or complement, given that the component that subscribes is not necesarily the component that receive notifications) to send initial notifications, it has been suggested to include information in the response to subscription creation operation, e.g. `matchedEntitiesIds` or a count of matched entities.

**Comment by fgalan** [comment]
Another alternative could be to use a HTTP header to "mark" the intiial notification.

**Comment by fgalan** [comment]
federation.md document now includes a paragraph regarding potential problems with the initial notification. That paragraph should be adapted once this issue gets completed, in order to mention the possibility of using this mechanism.

**Comment by fgalan** [comment]
initial_notification.md#additional-considerations should be adjusted once this issue gets completed

**Comment by fgalan** [comment]
This issue has been asigned to @SourabhChourasiya-NEC

Thanks!

**Comment by fgalan** [comment]
With regards to the specific option to avoid the initial notificaiton, what about?

```
POST /v2/subscription?options=skipInitialNotification
PATCH /v2/subscription/<id>?options=skipInitialNotification
```

`skipInitialNotification` is a bit long... @SourabhChourasiya-NEC (or anybody :) if you have some alternative name, please go ahead.

**Comment by fgalan** [comment]
> According to my understanding of code flow, initial notification is sent only once at subscription's creation time. Status of subscription can be changed from inactive to active but in this case also initial notification isn't triggered. So, changed are not required in PATCH request.

Note that the mongoUpdateSubscription logic can also send an initial notification https://github.com/telefonicaid/fiware-orion/blob/master/src/lib/mongoBackend/mongoUpdateSubscription.cpp#L905. Maybe the code is reused in both create/update operations so you would only touch the code in a single point but from the point of view of functional .test, both cases (create and update) should be covered.

> It is my suggestion that if we can add a optional field (e.g. initNotify="true") for initial notification in subscription's create payload, like status field. Its default value would be "true". If user doesn't want an initial notification, user can provide the field value "false" and no initial notification will be send.

We already has a way of specifing on/off aspects in NGSIv2 based in the  `options` URI parameter so I prefer to use it instead of inventing a new one. Default (if no such URI opetion is used) should be sending initial notification, to keep backward compatibility. Thus, `options=initNotify` will not work, but maybe `options=skipInitNotify`. What do you think?

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,

I do agree with you. But I have checked the functionality for subscription create/update. According to the code flow as we update the status from inactive to active through PATCH Subscription, notification should be sent to the consumer. But no notification is sent to consumer, it waits for next entity update for sending the notification.

As mentioned in the document of [initial notification](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/user/initial_notification.md) 

> However, there is a special case of notification, named initial notification, which is sent synchronous to the subscription creation (or update) transaction. This notification includes all the entities covered by the subscription being created (or updated).

According to this subscription update transaction, an initial notification should be send in case of creation and status updation as well.But when status is updated from inactive to active, initial notification is never sent. And I think it is bug and we should fix it.

And 

> We already has a way of specifing on/off aspects in NGSIv2 based in the options URI parameter so I prefer to use it instead of inventing a new one. Default (if no such URI opetion is used) should be sending initial notification, to keep backward compatibility. Thus, options=initNotify will not work, but maybe options=skipInitNotify. What do you think?

Yes Sir, we will use the URI parameter `options` for the same.

Thanks

**Comment by fgalan** [comment]
> According to this subscription update transaction, an initial notification should be send in case of creation and status updation as well.But when status is updated from inactive to active, initial notification is never sent. And I think it is bug and we should fix it.

`status` is not the only field you can update with PATCH on the subscription. What about if you PATCH changing the `subject` or `notification` fields?

**Comment by SourabhChourasiya-NEC** [comment]
Hi @fgalan Sir,
I have checked PATCH request for subject and notification field, but after updating one of the field or both, the initial notification doesn't trigger in this case also.
However, I have started working for configurable initial notification by URI parameter, and update you in some time.

Thank you.

**Comment by fgalan** [comment]
If you look for "initial notification" in existing .test you will find several ones related with initial notification on update. However, let's put appart that by the moment if you want and focus on the implementation.

Thanks!

**Comment by SourabhChourasiya-NEC** [comment]
fixed in #3397
